packages/api/tests/endpoints/collections/create-collection.js

```diff
@@ -291,3 +321,60 @@ test('POST with non-matching granuleId regex returns 400 bad request response',
   t.is(res.status, 400);
   t.true(res.body.message.includes('granuleId "badregex" cannot validate "filename"'));
 });
+
+test('post() does not write to the database if writing to Dynamo fails', async (t) => {
+  const { dbClient } = t.context;
+
+  const collection = fakeCollectionFactory();
+
+  const fakeCollectionsModel = {
+    exists: () => false,
+    create: () => {
+      throw new Error('something bad');
+    },
+  };
+
+  const expressRequest = {
+    body: collection,
+    testContext: {
+      dbClient,
+      collectionsModel: fakeCollectionsModel,
+    },
+  };
+
+  const response = buildFakeExpressResponse();
+
+  await post(expressRequest, response);
```

Comment:
One concern I have about this testing approach is that we're now relying on our internal knowledge of how an Express request is structured, which could break

Reply:
Unless there's some other reason to use this approach that I'm not recognizing, I'd probably prefer to use supertest and keep things consistent

Reply:
The reason is that I need to be able to simulate a Dynamo write failure and an RDS write failure, which injecting the clients allows me to do.

Reply:
Presumably mocking could achieve the same thing? I understand the appeal of direct "dependency injection" but worry a little when it relies on the internals of something like Express

Reply:
It probably could but, between the possibilities of the Express interface changing or the internals of our code changing (and breaking our tests), the latter seems much more likely. A change to Express's interface would also break a lot of our code. I'm also not sure that an Express request is really the internals of Express when it's documented as part of the library's API: https://expressjs.com/en/4x/api.html#req

Reply:
I'm not sure that I follow. The request structure is a documented part of the API but the API can change. Of course a future version of Express that changes its API could break all sorts of things but it seems like we increase that chance with this testing approach. 

Between inserting things into the interface of a package we don't manage and mocking the code that we are in control of, I prefer the latter.

I realize that this may seem pedantic, but I'm still not grasping why this approach is preferable to mocking given the above concerns. I'm happy to be convinced that I'm wrong if you want to discuss offline.

Reply:
I’d be interested in getting @nemreid and @Jkovarik’s opinions.